integrations/googledrive/src/client.ts

```diff
@@ -0,0 +1,313 @@
+  }
+
+  public async deleteFile(id: string) {
+    await this._googleClient.files.delete({
+      fileId: id,
+    })
+  }
+
+  public async uploadFileData({ id: fileId, url, mimeType }: UploadFileDataArgs) {
+    const downloadBpFileResp = await axios.get<Stream>(url, {
+      responseType: 'stream',
+    })
+    const downloadStream = downloadBpFileResp.data
+
+    await this._googleClient.files.update({
+      fileId,
+      media: {
+        body: downloadStream,
+        mimeType,
+      },
+    })
+  }
+
+  public async downloadFileData({ id: fileId, index }: DownloadFileDataArgs): Promise<DownloadFileDataOutput> {
+    // TODO: Must use export in case of a Google Workspace document.
+    // TODO: Need to map Google Workspace MIME types to the File's API supported MIME types
+    const file = await this._fetchFile(fileId)
+    if (file.type !== 'normal') {
+      throw new RuntimeError(`Attempted to download a file of type ${file.type}`)
+    }
+
+    const { mimeType: contentType, size } = file
+    const fileDownloadResponse = await this._googleClient.files.get(
+      {
+        fileId,
+        alt: 'media',
+      },
+      {
+        responseType: 'stream',
+      }
+    )
+    const fileDownloadStream = fileDownloadResponse.data
+    const { file: bpFile } = await this._client.upsertFile({
+      key: fileId,
+      size,
+      contentType,
+      index,
+    })
+    const { id: bpFileId, uploadUrl } = bpFile
+
+    const headers = {
+      'Content-Type': contentType,
+      'Content-Length': size,
+    }
+    await axios
+      .put(uploadUrl, fileDownloadStream, {
+        maxBodyLength: size,
+        headers,
+      })
+      .catch((reason: AxiosError) => {
+        throw new RuntimeError(`Error uploading file to botpress file API: ${reason}`)
+      })
+
+    return {
+      bpFileId,
+    }
+  }
+
+  /**
+   * Removes internal fields and adds computed attributes
+   */
+  private async _getCompleteFileFromBaseFile(file: BaseNormalFile, optionalFilesCache?: FilesCache): Promise<File> {
+    const genericFile = convertNormalFileToGeneric(file)
+    const { filesCache } = await this._addParentsToFilesCache(genericFile, optionalFilesCache)
+    const { id, mimeType, name, parentId, size } = file
+    return {
+      id,
+      mimeType,
+      name,
+      parentId,
+      size,
+      path: this._getFilePath(id, filesCache),
+    }
+  }
+
+  /**
+   * Removes internal fields and adds computed attributes
+   */
+  private async _getCompleteFolderFromBaseFolder(
+    file: BaseFolderFile,
+    optionalFilesCache?: FilesCache
+  ): Promise<Folder> {
+    const genericFile = convertFolderFileToGeneric(file)
+    const { filesCache } = await this._addParentsToFilesCache(genericFile, optionalFilesCache)
+    const { id, mimeType, name, parentId } = file
+    return {
+      id,
+      mimeType,
+      name,
+      parentId,
+      path: this._getFilePath(id, filesCache),
+    }
+  }
+
+  private async _fetchFilesAndUpdateCache({
```

Comment:
**chore:** This function is used twice. Both references get the output IDs and reconstruct the files using the cache... Why not return the whole files instead?

Reply:
I forgot to refactor after I added the type that enables discrimination between normal files and folders. I returned the IDs only and did the validation after.